File: server/app/domain/games.py
from dataclasses import dataclass
from enum import StrEnum


class GameCode(StrEnum):
    DOUBLE_COLOR_BALL = "ssq"
    SUPER_LOTTO = "dlt"


class GenerateMode(StrEnum):
    RANDOM = "random"
    EXCLUDE_HISTORY = "exclude_history"
    MODEL_PREDICTION = "model_prediction"


class ComplexPreference(StrEnum):
    SINGLE = "single"
    CONSERVATIVE = "conservative"
    BALANCED = "balanced"
    AGGRESSIVE = "aggressive"


@dataclass(frozen=True)
class GameDefinition:
    code: GameCode
    name: str
    primary_label: str
    secondary_label: str
    primary_range: range
    secondary_range: range
    primary_pick: int
    secondary_pick: int
    ticket_price: int
    base_first_prize: int
    rule_source_url: str
    draw_source_url: str


GAME_DEFINITIONS: dict[GameCode, GameDefinition] = {
    GameCode.DOUBLE_COLOR_BALL: GameDefinition(
        code=GameCode.DOUBLE_COLOR_BALL,
        name="Double Color Ball",
        primary_label="red_balls",
        secondary_label="blue_balls",
        primary_range=range(1, 34),
        secondary_range=range(1, 17),
        primary_pick=6,
        secondary_pick=1,
        ticket_price=2,
        base_first_prize=5_000_000,
        rule_source_url="https://www.cwl.gov.cn/c/2018-10-12/417937.shtml",
        draw_source_url=(
            "https://www.cwl.gov.cn/cwl_admin/front/cwlkj/search/kjxx/"
            "findDrawNotice?name=ssq"
        ),
    ),
    GameCode.SUPER_LOTTO: GameDefinition(
        code=GameCode.SUPER_LOTTO,
        name="Super Lotto",
        primary_label="front_zone",
        secondary_label="back_zone",
        primary_range=range(1, 36),
        secondary_range=range(1, 13),
        primary_pick=5,
        secondary_pick=2,
        ticket_price=2,
        base_first_prize=10_000_000,
        rule_source_url="https://m.lottery.gov.cn/ksjz/m/yxgz_dlt/",
        draw_source_url=(
            "https://webapi.sporttery.cn/gateway/lottery/"
            "getHistoryPageListV1.qry?gameNo=85"
        ),
    ),
}


def get_game_definition(game_code: GameCode | str) -> GameDefinition:
    return GAME_DEFINITIONS[GameCode(game_code)]

File: packages/domain/src/games.ts
import { ComplexPreference, GameCode, type GameDefinition } from "@lottery/contracts";

export const gameDefinitions: Record<GameCode, GameDefinition> = {
  [GameCode.DOUBLE_COLOR_BALL]: {
    code: GameCode.DOUBLE_COLOR_BALL,
    name: "双色球",
    primaryLabel: "redBalls",
    secondaryLabel: "blueBalls",
    primaryRange: [1, 33],
    secondaryRange: [1, 16],
    primaryPick: 6,
    secondaryPick: 1,
    ticketPrice: 2,
    baseFirstPrize: 5_000_000,
    ruleSourceUrl: "https://www.cwl.gov.cn/c/2018-10-12/417937.shtml",
    drawSourceUrl: "https://www.cwl.gov.cn/cwl_admin/front/cwlkj/search/kjxx/findDrawNotice?name=ssq",
  },
  [GameCode.SUPER_LOTTO]: {
    code: GameCode.SUPER_LOTTO,
    name: "大乐透",
    primaryLabel: "frontZone",
    secondaryLabel: "backZone",
    primaryRange: [1, 35],
    secondaryRange: [1, 12],
    primaryPick: 5,
    secondaryPick: 2,
    ticketPrice: 2,
    baseFirstPrize: 10_000_000,
    ruleSourceUrl: "https://m.lottery.gov.cn/ksjz/m/yxgz_dlt/",
    drawSourceUrl: "https://webapi.sporttery.cn/gateway/lottery/getHistoryPageListV1.qry?gameNo=85",
  },
};

export function getGameDefinition(gameCode: GameCode): GameDefinition {
  return gameDefinitions[gameCode];
}

export function listInclusiveRange([min, max]: [number, number]): number[] {
  return Array.from({ length: max - min + 1 }, (_, index) => min + index);
}

export function countRanges(gameCode: GameCode, preference: ComplexPreference): [[number, number], [number, number]] {
  if (gameCode === GameCode.DOUBLE_COLOR_BALL) {
    const ranges: Record<ComplexPreference, [[number, number], [number, number]]> = {
      [ComplexPreference.SINGLE]: [[6, 6], [1, 1]],
      [ComplexPreference.CONSERVATIVE]: [[7, 8], [1, 2]],
      [ComplexPreference.BALANCED]: [[8, 10], [2, 4]],
      [ComplexPreference.AGGRESSIVE]: [[10, 12], [3, 6]],
    };
    return ranges[preference];
  }
  const ranges: Record<ComplexPreference, [[number, number], [number, number]]> = {
    [ComplexPreference.SINGLE]: [[5, 5], [2, 2]],
    [ComplexPreference.CONSERVATIVE]: [[6, 7], [2, 3]],
    [ComplexPreference.BALANCED]: [[7, 9], [3, 5]],
    [ComplexPreference.AGGRESSIVE]: [[9, 11], [4, 7]],
  };
  return ranges[preference];
}
